### PR TITLE
Feature: Federate browse page

### PR DESCRIPTION
**Pull request by syphax-bouazzouni**
### Require 
* https://github.com/ontoportal-lirmm/ontologies_api_ruby_client/pull/17

### Context
See https://github.com/ontoportal/ontoportal-project/issues/38

### Changes 
* Show in the browse ontologies from external portals 
image

* Have custom styles depending on the origin ontology 
image

* Redirect to the external portal for federated ontologies 
image

* Federate filters  (categories, groups, languages, status, format, search) and sort by.

image

image

* Add an error message to inform when of the federated portals is down or returning an error
image

* Add configuration samples, using public API keys, the federation is enable for an external portal if the API URL and Key are provided 
```ruby
# Federation configuration
$PORTALS_INSTANCES = [
  {
    name: 'AgroPortal',
    api: 'https://data.agroportal.lirmm.fr',
    ui: 'https://agroportal.lirmm.fr/',
    color: '#349696',
    apikey: 'XXX',
    'light-color': '#F1F6FA',
  },
  {
    name: 'BioPortal',
    ui: 'https://bioportal.bioontology.org/',
    api: 'https://data.bioontology.org/',
    apikey: 'XXX',
    color: '#234979',
    'light-color': '#E9F2FA',
  },
  {
    name: 'SIFR BioPortal',
    ui: 'https://bioportal.lirmm.fr/',
    color: '#74a9cb',
  }
]
```

- Handle the merging of the same ontology from different portals (https://github.com/ontoportal-lirmm/bioportal_web_ui/pull/621/commits/f384920cf9defb727ae6158b0a411246a50f6ba0) 
image

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -124,7 +124,7 @@ def domain_ontology_set
       subdomain = host_parts[0].downcase
 
       slices = LinkedData::Client::Models::Slice.all
-      slices_acronyms = slices.map {|s| s.acronym}
+      slices_acronyms = slices.map {|s| s.acronym} rescue binding.pry
```
Comment (syphax-bouazzouni) [inline app/controllers/application_controller.rb:127]: to remove

File: config/initializers/ontologies_api_client.rb [hunk 2]
```diff
@@ -7,4 +7,5 @@
   config.debug_client = $DEBUG_RUBY_CLIENT || false
   config.debug_client_keys = $DEBUG_RUBY_CLIENT_KEYS || []
   config.apikey = $API_KEY
+  config.federated_portals = $FEDERATED_PORTALS || []
```
Comment (syphax-bouazzouni) [inline config/initializers/ontologies_api_client.rb:10]: change to `$FEDERATED_PORTALS || {}`

File: app/components/ontology_browse_card_component/ontology_browse_card_component.html.haml [hunk 3]
```diff
@@ -3,9 +3,11 @@
     .d-flex
       .browse-ontology-description
         .browse-ontology-title-bar
-          %a.browse-ontology-title{:href => "/ontologies/#{ontology[:acronym]}", data: {'turbo': 'false'}}
+          %a.browse-ontology-title{:href => onto_link, data: {'turbo': 'false'} , style: style_text}
             = ontology[:name]+" ("+ontology[:acronym]+")"
             = private_ontology_icon(ontology[:private])
+            - if external_ontology?
+              = render Display::InfoTooltipComponent.new(text: "Federated ontology from #{onto_link}", icon: 'external-link.svg')
```
Comment (Bilelkihal) [inline app/components/ontology_browse_card_component/ontology_browse_card_component.html.haml:10]: I don't really like this 'Federated' word, I don't think the regular user will understand what that is (To be discussed in Rome)

File: app/controllers/concerns/submission_filter.rb [hunk 4]
```diff
@@ -67,49 +58,36 @@ def ontologies_filter_url(filters, page: 1, count: false)
...
-      groups: groups, categories: categories,
-      formats: formats
-    )
-
-  end
-
   def filter_using_data(ontologies, query:, status:, show_views:, private_only:, languages:, page_size:, formality_level:, is_of_type:, groups:, categories:, formats:)
     submissions = LinkedData::Client::Models::OntologySubmission.all(include: BROWSE_ATTRIBUTES.join(','), also_include_views: true, display_links: false, display_context: false)
+
+    submissions = submissions.map { |x| [x[:ontology][:id], x] }.to_h rescue binding.pry
```
Comment (Bilelkihal) [inline app/controllers/concerns/submission_filter.rb:64]: To be deleted also

**Review by Bilelkihal — CHANGES_REQUESTED** [review]

File: app/components/ontology_browse_card_component/ontology_browse_card_component.html.haml [hunk 5]
```diff
@@ -3,9 +3,11 @@
     .d-flex
       .browse-ontology-description
         .browse-ontology-title-bar
-          %a.browse-ontology-title{:href => "/ontologies/#{ontology[:acronym]}", data: {'turbo': 'false'}}
+          %a.browse-ontology-title{:href => onto_link, data: {'turbo': 'false'} , style: style_text}
             = ontology[:name]+" ("+ontology[:acronym]+")"
             = private_ontology_icon(ontology[:private])
+            - if external_ontology?
```
[the remark this answers is not in the document]
Comment (syphax-bouazzouni) [reply app/components/ontology_browse_card_component/ontology_browse_card_component.html.haml:9]: can't do that easily, as there is no way to change its color, without creating a custom `css` for each portal.

**Comment by syphax-bouazzouni** [comment]
Deployed in https://testportal.lirmm.fr/ontologies

**Pull request merged**